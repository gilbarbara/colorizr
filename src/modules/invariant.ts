export function invariant(condition: boolean, message: string): asserts condition {
  if (condition) {
    return;
  }

  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  const error = !message
    ? new Error(
        'Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.',
      )
    : new Error(message);

  error.name = 'colorizr';

  throw error;
}
