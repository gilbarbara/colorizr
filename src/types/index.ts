/* eslint-disable typescript-sort-keys/interface */
export interface Analysis {
  brightnessDifference: number;
  colorDifference: number;
  compliant: number;
  contrast: number;
  largeAA: boolean;
  largeAAA: boolean;
  normalAA: boolean;
  normalAAA: boolean;
}

export interface Colors {
  hex: string;
  hsl: HSL;
  rgb: RGB;
}

export type ColorTypes = 'hex' | 'hsl' | 'rgb';
export type ColorModels = 'hsl' | 'rgb';

export interface FormatOptions {
  alpha?: number;
  model?: ColorModels;
}

export interface HSL {
  h: number;
  s: number;
  l: number;
}

export interface Options {
  model?: ColorModels;
}

export interface PaletteOptions {
  lightness?: number;
  saturation?: number;
  size?: number;
  type?: string;
}

export type PlainObject = Record<string, any>;

type ReturnModel<T> = T extends 'rgb' ? RGB : HSL;

export type Return<T> = T extends 'rgb' | 'hsl' ? ReturnModel<T> : string;

export interface RGB {
  r: number;
  g: number;
  b: number;
}

export type RGBArray = [number, number, number];

export type Scheme =
  | 'analogous'
  | 'complementary'
  | 'rectangle'
  | 'split'
  | 'split-complementary'
  | 'square'
  | 'tetradic'
  | 'triadic';
