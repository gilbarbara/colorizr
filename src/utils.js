// @flow
/**
 * Colorizr Utils
 * @module Utils
 */
/**
 * Parse math string expressions.
 *
 * @function
 * @param {string} input
 *
 * @returns {number}
 */
export function expr(input: string): number {
  const chars = input.split('');
  const n = [];
  const op = [];

  let parsed;
  let index = 0;
  let last = true;

  n[index] = '';

  // Parse the string
  for (let c = 0; c < chars.length; c++) {
    if (isNaN(parseInt(chars[c], 10)) && chars[c] !== '.' && !last) {
      op[index] = chars[c];
      index++;
      n[index] = '';
      last = true;
    }
    else {
      n[index] += chars[c];
      last = false;
    }
  }

  // Calculate the expression
  parsed = parseFloat(n[0]);
  for (let o = 0; o < op.length; o++) {
    const num = parseFloat(n[o + 1]);

    switch (op[o]) {
      case '+':
        parsed += num;
        break;
      case '-':
        parsed -= num;
        break;
      case '*':
        parsed *= num;
        break;
      case '/':
        parsed /= num;
        break;
      default:
        break;
    }
  }

  return parsed;
}

/**
 * Check if the input is an object
 *
 * @param {*} input
 * @returns {boolean}
 */
export function isPlainObject(input: any): boolean {
  if (!input) return false;

  const { toString } = Object.prototype;
  const prototype = Object.getPrototypeOf(input);

  return toString.call(input) === '[object Object]' && (prototype === null || prototype === Object.getPrototypeOf({}));
}

export function isRequired(name: string) {
  throw new Error(`${name || 'parameter'} is required`);
}

/**
 * Creates an object composed of the picked source properties.
 *
 * @param {Object} input
 * @param {Array} options
 *
 * @returns {Object}
 */
export function pick(input: Object, options: Array<string>): Object {
  if (!Array.isArray(options)) {
    throw new Error('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc, i) => {
      acc[i] = input[i];
      return acc;
    }, {});
}

/**
 * Round decimal numbers.
 *
 * @function
 * @param {number} input
 * @param {number} digits
 *
 * @returns {number}
 */
export function round(input: number, digits: number = 2): number {
  const factor = 10 ** digits;
  return Math.round(input * factor) / factor;
}

export function validateHex(input: string): boolean {
  return /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test(input);
}
