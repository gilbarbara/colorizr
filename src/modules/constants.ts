import { ColorKeysTuple, ColorModelKey } from '~/types';

export const COLOR_KEYS: Record<ColorModelKey, ColorKeysTuple> = {
  hsl: ['h', 's', 'l'],
  oklab: ['l', 'a', 'b'],
  oklch: ['l', 'c', 'h'],
  rgb: ['r', 'g', 'b'],
};
export const COLOR_MODELS: ColorModelKey[] = ['hsl', 'oklab', 'oklch', 'rgb'];
export const DEG2RAD = Math.PI / 180;
export const LAB_TO_LMS = {
  l: [0.3963377773761749, 0.2158037573099136],
  m: [-0.1055613458156586, -0.0638541728258133],
  s: [-0.0894841775298119, -1.2914855480194092],
};
export const LRGB_TO_LMS = {
  l: [0.4122214708, 0.5363325363, 0.0514459929],
  m: [0.2119034982, 0.6806995451, 0.1073969566],
  s: [0.0883024619, 0.2817188376, 0.6299787005],
};
export const LSM_TO_LAB = {
  l: [0.2104542553, 0.793617785, 0.0040720468],
  a: [1.9779984951, 2.428592205, 0.4505937099],
  b: [0.0259040371, 0.7827717662, 0.808675766],
};
export const LSM_TO_RGB = {
  r: [4.0767416360759583, -3.3077115392580629, 0.2309699031821043],
  g: [-1.2684379732850315, 2.6097573492876882, -0.341319376002657],
  b: [-0.0041960761386756, -0.7034186179359362, 1.7076146940746117],
};
export const PRECISION = 5;
export const RAD2DEG = 180 / Math.PI;

export const MESSAGES = {
  alpha: 'amount must be a number between 0 and 1',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputHex: 'input is required and must be a hex',
  inputNumber: 'input is required and must be a number',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  invalidCSS: 'invalid CSS string',
  options: 'invalid options',
  threshold: 'threshold must be a number between 0 and 255',
} as const;
