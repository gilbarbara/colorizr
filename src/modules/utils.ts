import { HSL, PlainObject, RGB, RGBArray } from '../types';

export const HSLKeys = ['h', 's', 'l'];
export const RGBKeys = ['r', 'g', 'b'];

/**
 * Constrain value into the range
 */
export function constrain(input: number, amount: number, range: number[], sign: string): number {
  invariant(arguments.length === 4, 'All parameters are required');

  const [min, max] = range;
  let value = expr(input + sign + amount);

  if (value < min) {
    value = min;
  } else if (value > max) {
    value = max;
  }

  return Math.abs(value);
}

/**
 * Constrain an angle
 */
export function constrainDegrees(input: number, amount: number): number {
  invariant(isNumber(input), 'input is required');

  let value = input + amount;

  if (value > 360) {
    value %= 360;
  }

  if (value < 0) {
    value += 360;
  }

  return Math.abs(value);
}

/**
 * Parse math string expressions
 */
export function expr(input: string): number {
  const chars = [...input];
  const n: string[] = [];
  const op: string[] = [];

  let parsed;
  let index = 0;
  let last = true;

  n[index] = '';

  // Parse the string
  for (const char of chars) {
    if (Number.isNaN(parseInt(char, 10)) && char !== '.' && !last) {
      op[index] = char;
      index++;
      n[index] = '';
      last = true;
    } else {
      n[index] += char;
      last = false;
    }
  }

  // Calculate the expression
  parsed = parseFloat(n[0]);

  for (const [o, element] of op.entries()) {
    const value = parseFloat(n[o + 1]);

    switch (element) {
      case '+':
        parsed += value;
        break;
      case '-':
        parsed -= value;
        break;
      case '*':
        parsed *= value;
        break;
      case '/':
        parsed /= value;
        break;
      default:
        break;
    }
  }

  return parsed;
}

export function invariant(condition: boolean, message: string): asserts condition {
  if (condition) {
    return;
  }

  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  let error;

  if (!message) {
    throw new Error(
      'Minified exception occurred; use the non-minified dev environment ' +
        'for the full error message and additional helpful warnings.',
    );
  } else {
    error = new Error(message);
  }

  error.name = 'colorizr';

  throw error;
}

/**
 * Check if an object contains HSL values
 */
export function isHSL(input: any): input is HSL {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(
      ([key, value]) => HSLKeys.includes(key) && value >= 0 && value <= (key === 'h' ? 360 : 100),
    )
  );
}

/**
 * Check if the input is a number and not NaN
 */
export function isNumber(input: any): input is number {
  return typeof input === 'number' && !Number.isNaN(input);
}

/**
 * Check if the input is an object
 */
export function isPlainObject(input: any): input is PlainObject {
  if (!input) {
    return false;
  }

  const { toString } = Object.prototype;
  const prototype = Object.getPrototypeOf(input);

  return (
    toString.call(input) === '[object Object]' &&
    (prototype === null || prototype === Object.getPrototypeOf({}))
  );
}

/**
 * Check if an object contains RGB values.
 */
export function isRGB(input: any): input is RGB {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(([key, value]) => RGBKeys.includes(key) && value >= 0 && value <= 255)
  );
}

/**
 * Check if an array contains RGB values.
 */
export function isRGBArray(input: any): input is RGBArray {
  return Array.isArray(input) && input.length === 3 && input.every(d => d >= 0 && d <= 255);
}

/**
 * Check if the input is a string
 */
export function isString(input: any): input is string {
  return typeof input === 'string';
}

/**
 * Limit values per type.
 */
export function limit(input: number, type: string): number {
  invariant(isNumber(input), 'Input is not a number');

  /* istanbul ignore else */
  if (RGBKeys.includes(type)) {
    return Math.max(Math.min(input, 255), 0);
  }

  if (['s', 'l'].includes(type)) {
    return Math.max(Math.min(input, 100), 0);
  }

  if (type === 'h') {
    return Math.max(Math.min(input, 360), 0);
  }

  throw new Error('Invalid type');
}

export const messages = {
  amount: 'amount must be a number',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  options: 'invalid options',
};

/**
 * Creates an object composed of the picked source properties.
 */
export function pick(input: PlainObject, options: string[]): PlainObject {
  if (!Array.isArray(options)) {
    throw new TypeError('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc: PlainObject, d) => {
      acc[d] = input[d];

      return acc;
    }, {});
}

/**
 * Round decimal numbers.
 */
export function round(input: number, digits = 2): number {
  const factor = 10 ** digits;

  return Math.round(input * factor) / factor;
}
