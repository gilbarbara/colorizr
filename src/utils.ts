import { HSL, RGB, PlainObject, RGBArray } from './types';

export const HSLKeys = ['h', 's', 'l'];
export const RGBKeys = ['r', 'g', 'b'];

/**
 * Constrain value into the range
 */
export function constrain(input: number, amount: number, range: number[], sign: string): number {
  invariant(arguments.length < 4, 'All parameters are required');

  const [min, max] = range;
  let value = expr(input + sign + amount);

  if (value < min) {
    value = min;
  } else if (value > max) {
    value = max;
  }

  return Math.abs(value);
}

/**
 * Constrain an angle
 */
export function constrainDegrees(input: number, amount: number): number {
  invariant(!isNumber(input), 'input is required');

  let value = input + amount;

  if (value > 360) {
    value %= 360;
  }

  if (value < 0) {
    value += 360;
  }

  return Math.abs(value);
}

/**
 * CSS named colors
 */
export const cssColors = {
  aliceblue: '#f0f8ff',
  antiquewhite: '#faebd7',
  aqua: '#00ffff',
  aquamarine: '#7fffd4',
  azure: '#f0ffff',
  beige: '#f5f5dc',
  bisque: '#ffe4c4',
  black: '#000000',
  blanchedalmond: '#ffebcd',
  blue: '#0000ff',
  blueviolet: '#8a2be2',
  brown: '#a52a2a',
  burlywood: '#deb887',
  cadetblue: '#5f9ea0',
  chartreuse: '#7fff00',
  chocolate: '#d2691e',
  coral: '#ff7f50',
  cornflowerblue: '#6495ed',
  cornsilk: '#fff8dc',
  crimson: '#dc143c',
  cyan: '#00ffff',
  darkblue: '#00008b',
  darkcyan: '#008b8b',
  darkgoldenrod: '#b8860b',
  darkgray: '#a9a9a9',
  darkgrey: '#a9a9a9',
  darkgreen: '#006400',
  darkkhaki: '#bdb76b',
  darkmagenta: '#8b008b',
  darkolivegreen: '#556b2f',
  darkorange: '#ff8c00',
  darkorchid: '#9932cc',
  darkred: '#8b0000',
  darksalmon: '#e9967a',
  darkseagreen: '#8fbc8f',
  darkslateblue: '#483d8b',
  darkslategray: '#2f4f4f',
  darkslategrey: '#2f4f4f',
  darkturquoise: '#00ced1',
  darkviolet: '#9400d3',
  deeppink: '#ff1493',
  deepskyblue: '#00bfff',
  dimgray: '#696969',
  dimgrey: '#696969',
  dodgerblue: '#1e90ff',
  firebrick: '#b22222',
  floralwhite: '#fffaf0',
  forestgreen: '#228b22',
  fuchsia: '#ff00ff',
  gainsboro: '#dcdcdc',
  ghostwhite: '#f8f8ff',
  gold: '#ffd700',
  goldenrod: '#daa520',
  gray: '#808080',
  grey: '#808080',
  green: '#008000',
  greenyellow: '#adff2f',
  honeydew: '#f0fff0',
  hotpink: '#ff69b4',
  indianred: '#cd5c5c',
  indigo: '#4b0082',
  ivory: '#fffff0',
  khaki: '#f0e68c',
  lavender: '#e6e6fa',
  lavenderblush: '#fff0f5',
  lawngreen: '#7cfc00',
  lemonchiffon: '#fffacd',
  lightblue: '#add8e6',
  lightcoral: '#f08080',
  lightcyan: '#e0ffff',
  lightgoldenrodyellow: '#FAFAD2',
  lightgray: '#d3d3d3',
  lightgrey: '#d3d3d3',
  lightgreen: '#90ee90',
  lightpink: '#ffb6c1',
  lightsalmon: '#ffa07a',
  lightseagreen: '#20b2aa',
  lightskyblue: '#87cefa',
  lightslategray: '#778899',
  lightslategrey: '#778899',
  lightsteelblue: '#b0c4de',
  lightyellow: '#ffffe0',
  lime: '#00ff00',
  limegreen: '#32cd32',
  linen: '#faf0e6',
  magenta: '#ff00ff',
  maroon: '#800000',
  mediumaquamarine: '#66cdaa',
  mediumblue: '#0000cd',
  mediumorchid: '#ba55d3',
  mediumpurple: '#9370d8',
  mediumseagreen: '#3cb371',
  mediumslateblue: '#7b68ee',
  mediumspringgreen: '#00fa9a',
  mediumturquoise: '#48d1cc',
  mediumvioletred: '#c71585',
  midnightblue: '#191970',
  mintcream: '#f5fffa',
  mistyrose: '#ffe4e1',
  moccasin: '#ffe4b5',
  navajowhite: '#ffdead',
  navy: '#000080',
  oldlace: '#fdf5e6',
  olive: '#808000',
  olivedrab: '#6b8e23',
  orange: '#ffa500',
  orangered: '#ff4500',
  orchid: '#da70d6',
  palegoldenrod: '#eee8aa',
  palegreen: '#98fb98',
  paleturquoise: '#afeeee',
  palevioletred: '#d87093',
  papayawhip: '#ffefd5',
  peachpuff: '#ffdab9',
  peru: '#cd853f',
  pink: '#ffc0cb',
  plum: '#dda0dd',
  powderblue: '#b0e0e6',
  purple: '#800080',
  red: '#ff0000',
  rosybrown: '#bc8f8f',
  royalblue: '#4169e1',
  saddlebrown: '#8b4513',
  salmon: '#fa8072',
  sandybrown: '#f4a460',
  seagreen: '#2e8b57',
  seashell: '#fff5ee',
  sienna: '#a0522d',
  silver: '#c0c0c0',
  skyblue: '#87ceeb',
  slateblue: '#6a5acd',
  slategray: '#708090',
  slategrey: '#708090',
  snow: '#fffafa',
  springgreen: '#00ff7f',
  steelblue: '#4682b4',
  tan: '#d2b48c',
  teal: '#008080',
  thistle: '#d8bfd8',
  tomato: '#ff6347',
  turquoise: '#40e0d0',
  violet: '#ee82ee',
  wheat: '#f5deb3',
  white: '#ffffff',
  whitesmoke: '#f5f5f5',
  yellow: '#ffff00',
  yellowgreen: '#9acd32',
};

/**
 * Parse math string expressions
 */
export function expr(input: string): number {
  const chars = input.split('');
  const n: string[] = [];
  const op: string[] = [];

  let parsed;
  let index = 0;
  let last = true;

  n[index] = '';

  // Parse the string
  for (const char of chars) {
    if (Number.isNaN(parseInt(char, 10)) && char !== '.' && !last) {
      op[index] = char;
      index++;
      n[index] = '';
      last = true;
    } else {
      n[index] += char;
      last = false;
    }
  }

  // Calculate the expression
  parsed = parseFloat(n[0]);
  for (let o = 0; o < op.length; o++) {
    const num = parseFloat(n[o + 1]);

    switch (op[o]) {
      case '+':
        parsed += num;
        break;
      case '-':
        parsed -= num;
        break;
      case '*':
        parsed *= num;
        break;
      case '/':
        parsed /= num;
        break;
      default:
        break;
    }
  }

  return parsed;
}

export function invariant(
  condition: boolean,
  message: string,
  a?: any,
  b?: any,
  c?: any,
  d?: any,
  e?: any,
  f?: any,
): void {
  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  if (condition) {
    let error;

    if (!message) {
      error = new Error(
        'Minified exception occurred; use the non-minified dev environment ' +
          'for the full error message and additional helpful warnings.',
      );
    } else {
      const args = [a, b, c, d, e, f];
      let argIndex = 0;
      error = new Error(message.replace(/%s/g, () => args[argIndex++]));
      error.name = 'colorizr';
    }

    throw error;
  }
}

/**
 * Check if an object contains HSL values
 */
export function isHSL(input: any): input is HSL {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(
      ([key, value]) =>
        HSLKeys.indexOf(key) >= 0 && value >= 0 && value <= (key === 'h' ? 360 : 100),
    )
  );
}

/**
 * Check if the input is a number and not NaN
 */
export function isNumber(input: any): input is number {
  return typeof input === 'number' && !Number.isNaN(input);
}

/**
 * Check if the input is an object
 */
export function isPlainObject(input: any): input is PlainObject {
  if (!input) {
    return false;
  }

  const { toString } = Object.prototype;
  const prototype = Object.getPrototypeOf(input);

  return (
    toString.call(input) === '[object Object]' &&
    (prototype === null || prototype === Object.getPrototypeOf({}))
  );
}

/**
 * Check if an object contains RGB values.
 */
export function isRGB(input: any): input is RGB {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(([key, value]) => RGBKeys.indexOf(key) >= 0 && value >= 0 && value <= 255)
  );
}

/**
 * Check if an array contains RGB values.
 */
export function isRGBArray(input: any): input is RGBArray {
  return Array.isArray(input) && input.length === 3 && input.every(d => d >= 0 && d <= 255);
}

/**
 * Check if the input is a string
 */
export function isString(input: any): input is string {
  return typeof input === 'string';
}

/**
 * Limit values per type.
 */
export function limit(input: number, type: string): number {
  invariant(!isNumber(input), 'Input is not a number');

  /* istanbul ignore else */
  if (RGBKeys.indexOf(type) >= 0) {
    return Math.max(Math.min(input, 255), 0);
  }

  if (['s', 'l'].indexOf(type) >= 0) {
    return Math.max(Math.min(input, 100), 0);
  }

  if (type === 'h') {
    return Math.max(Math.min(input, 360), 0);
  }

  throw new Error('Invalid type');
}

export const messages = {
  amount: 'amount must be a number',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  options: 'invalid options',
};

/**
 * Creates an object composed of the picked source properties.
 */
export function pick(input: PlainObject, options: string[]): PlainObject {
  if (!Array.isArray(options)) {
    throw new Error('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc: PlainObject, d) => {
      acc[d] = input[d];
      return acc;
    }, {});
}

/**
 * Round decimal numbers.
 */
export function round(input: number, digits = 2): number {
  const factor = 10 ** digits;
  return Math.round(input * factor) / factor;
}
