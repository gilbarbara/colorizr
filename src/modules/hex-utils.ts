import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { round } from '~/modules/utils';
import { isHex, isNumber } from '~/modules/validators';

import { Alpha } from '~/types';

/**
 * Add an alpha value to a hex string
 */
export function addAlphaToHex(input: string, alpha: Alpha): string {
  invariant(isHex(input), MESSAGES.inputHex);
  invariant(isNumber(alpha), MESSAGES.inputNumber);

  if (alpha >= 1) {
    return removeAlphaFromHex(input);
  }

  return `${removeAlphaFromHex(input)}${convertAlphaToHex(alpha)}`;
}

/**
 * Convert an alpha value to a hex value.
 */
export function convertAlphaToHex(input: Alpha): string {
  invariant(isNumber(input), MESSAGES.inputNumber);

  let alpha = input;

  if (input > 1) {
    alpha /= 100;
  }

  return Math.round(alpha * 255)
    .toString(16)
    .padStart(2, '0');
}

/**
 * Extract the alpha value from a hex string
 */
export function extractAlphaFromHex(input: string): number {
  invariant(isHex(input), MESSAGES.inputString);

  const alpha = input.substring(7, 9);

  if (!alpha) {
    return 1;
  }

  return round(parseInt(alpha, 16) / 255);
}

export function hexadecimalToNumber(input: string) {
  return round(parseInt(input, 16));
}

/**
 * Remove the alpha value from a hex string
 */
export function removeAlphaFromHex(input: string) {
  invariant(isHex(input), MESSAGES.inputHex);

  if (input.length === 5) {
    return input.substring(0, 4);
  }

  return input.substring(0, 7);
}
