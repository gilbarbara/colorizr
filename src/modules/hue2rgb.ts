import { invariant, isNumber, round } from './utils';

/**
 * Convert hue to RGB using chroma and median point
 */
export default function hue2rgb(point: number, chroma: number, h: number): number {
  invariant(isNumber(point) && isNumber(chroma) && isNumber(h), 'point, chroma and h are required');
  let hue = h;

  if (hue < 0) {
    hue += 1;
  }

  if (hue > 1) {
    hue -= 1;
  }

  if (hue < 1 / 6) {
    return round(point + (chroma - point) * 6 * hue, 4);
  }

  if (hue < 1 / 2) {
    return round(chroma, 4);
  }

  if (hue < 2 / 3) {
    return round(point + (chroma - point) * (2 / 3 - hue) * 6, 4);
  }

  return round(point, 4);
}
