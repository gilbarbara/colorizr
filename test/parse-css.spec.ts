import parseCSS from 'parse-css';

import { messages } from 'modules/utils';

import { ColorTypes } from 'types';

describe('parseCSS', () => {
  describe.each([['hex'], ['hsl'], ['rgb']] as ColorTypes[][])('with output %p', type => {
    it.each([
      ['#ff0044'],
      ['AliceBlue'],
      ['greenyellow'],
      ['rgb(255, 255, 0)'],
      ['rgba(255, 127, 12, 0.5)'],
      ['hsl(255,80%,20%)'],
      ['hsla(126,     57%,  62%,0.5)'],
      ['hsl(270 60% 70%)'],
      ['hsl(270 60%, 70% / .15)'],
    ])('%p should return properly', input => {
      expect(parseCSS(input, type)).toMatchSnapshot();
    });
  });

  it('should fail with invalid parameters', () => {
    // @ts-ignore
    expect(() => parseCSS([25, 56, 84])).toThrow(messages.inputString);
    // @ts-ignore
    expect(() => parseCSS({ h: 25, s: 56, l: 84 })).toThrow(messages.inputString);
    // @ts-ignore
    expect(() => parseCSS()).toThrow(messages.inputString);
    expect(() => parseCSS('rgb(255, 255)')).toThrow('invalid CSS string');
    expect(() => parseCSS('rgs(255, 255, 0)')).toThrow('invalid CSS string');
    expect(() => parseCSS('AliceGreen')).toThrow('invalid CSS string');
    expect(() => parseCSS('green-ish')).toThrow('invalid CSS string');
  });
});
