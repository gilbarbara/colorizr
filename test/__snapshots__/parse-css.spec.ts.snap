// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseCSS with output "hex" "#ff0044" should return properly 1`] = `"#ff0044"`;

exports[`parseCSS with output "hex" "AliceBlue" should return properly 1`] = `"#f0f8ff"`;

exports[`parseCSS with output "hex" "greenyellow" should return properly 1`] = `"#adff2f"`;

exports[`parseCSS with output "hex" "hsl(255,80%,20%)" should return properly 1`] = `"#1f0a5c"`;

exports[`parseCSS with output "hex" "hsl(270 60% 70%)" should return properly 1`] = `"#b385e0"`;

exports[`parseCSS with output "hex" "hsl(270 60%, 70% / .15)" should return properly 1`] = `"#b385e0"`;

exports[`parseCSS with output "hex" "hsla(126,     57%,  62%,0.5)" should return properly 1`] = `"#67d572"`;

exports[`parseCSS with output "hex" "rgb(255, 255, 0)" should return properly 1`] = `"#ffff00"`;

exports[`parseCSS with output "hex" "rgba(255, 127, 12, 0.5)" should return properly 1`] = `"#ff7f0c"`;

exports[`parseCSS with output "hsl" "#ff0044" should return properly 1`] = `
{
  "h": 344,
  "l": 50,
  "s": 100,
}
`;

exports[`parseCSS with output "hsl" "AliceBlue" should return properly 1`] = `
{
  "h": 208,
  "l": 97.06,
  "s": 100,
}
`;

exports[`parseCSS with output "hsl" "greenyellow" should return properly 1`] = `
{
  "h": 83.65,
  "l": 59.22,
  "s": 100,
}
`;

exports[`parseCSS with output "hsl" "hsl(255,80%,20%)" should return properly 1`] = `
{
  "h": 255,
  "l": 20,
  "s": 80,
}
`;

exports[`parseCSS with output "hsl" "hsl(270 60% 70%)" should return properly 1`] = `
{
  "h": 270,
  "l": 70,
  "s": 60,
}
`;

exports[`parseCSS with output "hsl" "hsl(270 60%, 70% / .15)" should return properly 1`] = `
{
  "h": 270,
  "l": 70,
  "s": 60,
}
`;

exports[`parseCSS with output "hsl" "hsla(126,     57%,  62%,0.5)" should return properly 1`] = `
{
  "h": 126,
  "l": 62,
  "s": 57,
}
`;

exports[`parseCSS with output "hsl" "rgb(255, 255, 0)" should return properly 1`] = `
{
  "h": 60,
  "l": 50,
  "s": 100,
}
`;

exports[`parseCSS with output "hsl" "rgba(255, 127, 12, 0.5)" should return properly 1`] = `
{
  "h": 28.4,
  "l": 52.35,
  "s": 100,
}
`;

exports[`parseCSS with output "rgb" "#ff0044" should return properly 1`] = `
{
  "b": 68,
  "g": 0,
  "r": 255,
}
`;

exports[`parseCSS with output "rgb" "AliceBlue" should return properly 1`] = `
{
  "b": 255,
  "g": 248,
  "r": 240,
}
`;

exports[`parseCSS with output "rgb" "greenyellow" should return properly 1`] = `
{
  "b": 47,
  "g": 255,
  "r": 173,
}
`;

exports[`parseCSS with output "rgb" "hsl(255,80%,20%)" should return properly 1`] = `
{
  "b": 92,
  "g": 10,
  "r": 31,
}
`;

exports[`parseCSS with output "rgb" "hsl(270 60% 70%)" should return properly 1`] = `
{
  "b": 224,
  "g": 133,
  "r": 179,
}
`;

exports[`parseCSS with output "rgb" "hsl(270 60%, 70% / .15)" should return properly 1`] = `
{
  "b": 224,
  "g": 133,
  "r": 179,
}
`;

exports[`parseCSS with output "rgb" "hsla(126,     57%,  62%,0.5)" should return properly 1`] = `
{
  "b": 114,
  "g": 213,
  "r": 103,
}
`;

exports[`parseCSS with output "rgb" "rgb(255, 255, 0)" should return properly 1`] = `
{
  "b": 0,
  "g": 255,
  "r": 255,
}
`;

exports[`parseCSS with output "rgb" "rgba(255, 127, 12, 0.5)" should return properly 1`] = `
{
  "b": 12,
  "g": 127,
  "r": 255,
}
`;
